### PR TITLE
Crysis Nanosuit

### PR DESCRIPTION
**Pull request by yoyobatty**
[Changelogs]: # (Your PR should contain a detailed changelog of notable changes, titled and categorized appropriately. This includes, new features, sprites, sounds, balance changes, admin tools, map edits, removals, big refactors, config changes, hosting changes and important fixes. An example changelog has been provided below for you to edit. If you need additional help, read https://github.com/tgstation/tgstation/wiki/Changelogs)

:cl: YoYoBatty
add: Adds the Crysis Nanosuit 
/:cl:

[why]: # (Please add a short description [two lines down] of why you think these changes would benefit the game. If you can't justify it in words, it might not be worth adding.)

Hey guys, I've been working for the past couple of months on this. I will update this and include more screenshots and information later on. 

So, what is it? 
Well it's the nanosuit from crysis.
What it do?
Super powerful armor/hardsuit that has 4 modes: Armor, cloak, speed and strength. Passive modes include toggleable night vision and zooming. Comes with a built-in disintegration implant that activates upon user death or request. 
Armor does exactly what it says it does and gives you a major buff to your helmet and suit armor + a shield which makes you immune from most attacks at the cost of suit energy. 
Cloak is very similar to the genetic mutation or ninja cloak mode.
Speed lets the user run faster than full speed while also ignoring any slowdowns.
Strength makes you a strong man and lets you over/onto tables but costs around 30 energy per use.
Is it OP?
Sorta. But not. It has a charge which self charges when not in use but is quickly depleted using certain modes. Armor drains energy at a rate of 5 + damage per hit. Cloak drains passively at once per process and on movement. Speed is just like cloak but only drains when you move. Strength gives you a slighter more robust krav maga that lets you deal increased damage with your fist to objects, doors and players and also increases the power of items in your hand. In addition to increasing your physical strength it gives you the ability to jump over objects (similar to jump boots but less overpowered and better code). 
Upon equipping the suit all your items are dropped below you are equipped with a non-droppable outfit which includes a syndie bowman headset that can't be emped, shock and fireproof gloves, slip proof boots, night vision goggles and mask. Everything has radiation proofing and completely seals the user from the outside world to prevent the operator from being infected by diseases. 

The suit also comes equipped with a series of sensors which constantly monitor user vitals.
There is also a built-in very minor medical injector which will activate when the user gets hit below 50 health, the chance being random and increases depending on how much damage was taken. The injector has 3 and regenerates every 20 seconds. 

Edit: This is up for testing and very possible merging.

https://www.youtube.com/watch?v=VIWBVZprdL8 short demo

### CODE REVIEW

**Comment by JohnGinnane** [comment]
@yoyobatty can you post screenshots and/or video?

**Comment by yoyobatty** [comment]
@JohnGinnane hey man, sorry just been busy with school and work, I'll get some screenshots and maybe a video going on everything. As for testing and functionality, I've probably compiled over 200 times extensively testing and rewriting code, so I'm hoping to have this out sometime this week.

As for it being a traitor item, this will cost 20tc and I plan on having it available as a special kit item in traitor uplinks, similar to Kanana or Bowling.

**Comment by JohnGinnane** [comment]
@yoyobatty no worries, don't rush yourself

**Comment by JohnMcLaime** [comment]
Strength mode should buff/speed up grabbing and increase the throwing power too.

**Comment by yoyobatty** [comment]
Hey @johnmclaime this is already a feature,  strength grants you a martial art that defaults you to have aggressive grab also it grants you the push immunity trait.

**Comment by JohnMcLaime** [comment]
What about throwing stuff?
Anyway this seems too good for 20 tc, maybe make a centcom announcement if someone buys one and then crew can order similar chinese knockoff suits from cargo too as a countermeasure.

Or make this nukie only for 40+ tc (since shielded hardsuit is 35).

**Comment by yoyobatty** [comment]
I am still working on getting thrown items to hit harder, essentially I just have to make them throw faster whenever the suit is in strength mode. One of the last things I need to complete as well is making firing guns and attacking disable cloak mode + adding overlays to the suit for specific modes.

I do want this item to be used in normal rounds too and I definitely feel as though it will require a bit of balancing. I like the announcement idea as well, maybe I will include the location of the individual and name of whoever puts it on.

File: hippiestation/code/modules/clothing/spacesuits/nanosuit.dm [hunk 1]
```diff
@@ -0,0 +1,934 @@
...
+	..()
+	qdel(src)
+
+/obj/item/device/radio/headset/syndicate/alt/nano
+	name = "\proper the nanosuit's bowman headset"
+	desc = "Operator communication headset. Property of CryNet Systems."
+	icon_state = "syndie_headset"
+	item_state = "syndie_headset"
+	subspace_transmission = FALSE
+	keyslot = new /obj/item/device/encryptionkey/binary
```
Comment (GuyonBroadway) [inline hippiestation/code/modules/clothing/spacesuits/nanosuit.dm:135]: Just to be clear, this is the key that lets you talk and listen to the AI Binary channel and not the one that lets you listen to all channels. Is this what you intended?

File: hippiestation/code/modules/clothing/spacesuits/nanosuit.dm [hunk 2]
```diff
@@ -0,0 +1,934 @@
...
+/obj/item/clothing/glasses/nano_goggles/AltClick(mob/user)
+	return
+
+/obj/item/clothing/glasses/nano_goggles/proc/nvgmode(mob/user, var/forced = FALSE)
+	on = !on
+	to_chat(user, "<span class='[forced ? "warning":"notice"]'>[forced ? "The goggles turn":"You turn the goggles"] [on ? "on":"off"][forced ? "!":"."]</span>")
+	if(ishuman(user))
+		var/mob/living/carbon/human/H = user
+		if(H.glasses == src)
+			switch(on)
```
Comment (ghost) [inline hippiestation/code/modules/clothing/spacesuits/nanosuit.dm]: could just have this as not a switch statement
if(on)
else

File: hippiestation/code/modules/clothing/spacesuits/nanosuit.dm [hunk 3]
```diff
@@ -0,0 +1,934 @@
...
+		sleep(recharge_delay)//recharges us at variable rate
+
+
+/obj/item/clothing/suit/space/hardsuit/nano/hit_reaction(mob/living/carbon/human/user, atom/movable/hitby, attack_text = "the attack", final_block_chance = 0, damage = 0, attack_type = MELEE_ATTACK)
+	var/obj/item/projectile/P = hitby
+	if(mode == "armor")
+		if(cell.charge > 0)
+			cell.use(CLAMP(hit_use + damage,1,cell.charge))
+			user.visible_message("<span class='danger'>[user]'s shields deflect [attack_text]!</span>")
+			return 1
```
Comment (GuyonBroadway) [inline hippiestation/code/modules/clothing/spacesuits/nanosuit.dm]: return TRUE

File: hippiestation/code/modules/clothing/spacesuits/nanosuit.dm [hunk 4]
```diff
@@ -0,0 +1,934 @@
...
+		else
+			return 0
```
Comment (GuyonBroadway) [inline hippiestation/code/modules/clothing/spacesuits/nanosuit.dm]: return FALSE

File: hippiestation/code/modules/clothing/spacesuits/nanosuit.dm [hunk 5]
```diff
@@ -0,0 +1,934 @@
...
+			var/datum/effect_system/spark_spread/s = new
+			s.set_up(1, 1, src)
+			s.start()
+	kill_cloak(user)
+	if(prob(damage*2.5) && user.health < 50 && current_charges > 0)
+		medical_cooldown = world.time + medical_delay
+		current_charges--
+		heal_nano(user)
+
+	return 0
```
Comment (GuyonBroadway) [inline hippiestation/code/modules/clothing/spacesuits/nanosuit.dm]: return FALSE

File: hippiestation/code/modules/clothing/spacesuits/nanosuit.dm [hunk 6]
```diff
@@ -0,0 +1,934 @@
...
+	if(A.grab_state >= GRAB_AGGRESSIVE)
+		D.grabbedby(A, 1)
+	else
+		A.start_pulling(D, 1)
+		if(A.pulling)
+			D.stop_pulling()
+			add_logs(A, D, "grabbed", addition="aggressively")
+			A.grab_state = GRAB_AGGRESSIVE //Instant aggressive grab
+
+	return 1
```
Comment (GuyonBroadway) [inline hippiestation/code/modules/clothing/spacesuits/nanosuit.dm]: return TRUE

File: hippiestation/code/modules/clothing/spacesuits/nanosuit.dm [hunk 7]
```diff
@@ -0,0 +1,934 @@
...
+							"<span class='userdanger'>[A] knocks you the fuck down!</span>")
+		else if(A.grab_state > GRAB_AGGRESSIVE)
+			var/atom/throw_target = get_edge_target_turf(D, A.dir)
+			if(!D.anchored)
+				D.throw_at(throw_target, rand(1,2), 7, A)
+			bonus_damage += 10
+			D.Knockdown(60)
+			D.visible_message("<span class='warning'>[A] knocks [D] the fuck out!!", \
+							"<span class='userdanger'>[A] knocks you the fuck out!!</span>")
+	return 1
```
Comment (GuyonBroadway) [inline hippiestation/code/modules/clothing/spacesuits/nanosuit.dm]: return TRUE

File: hippiestation/code/modules/clothing/spacesuits/nanosuit.dm [hunk 8]
```diff
@@ -0,0 +1,934 @@
...
+		D.visible_message("<span class='danger'>[A] has disarmed [D]!</span>", \
+							"<span class='userdanger'>[A] has disarmed [D]!</span>")
+		playsound(D, 'sound/weapons/thudswoosh.ogg', 50, 1, -1)
+		D.Knockdown(40)
+	else
+		D.visible_message("<span class='danger'>[A] attempted to disarm [D]!</span>", \
+							"<span class='userdanger'>[A] attempted to disarm [D]!</span>")
+		playsound(D, 'sound/weapons/punchmiss.ogg', 25, 1, -1)
+	add_logs(A, D, "disarmed with krav maga", "[I ? " removing \the [I]" : ""]")
+	return 1
```
Comment (GuyonBroadway) [inline hippiestation/code/modules/clothing/spacesuits/nanosuit.dm]: return TRUE

File: hippiestation/code/modules/clothing/spacesuits/nanosuit.dm [hunk 9]
```diff
@@ -0,0 +1,934 @@
...
+
+/atom/proc/attack_nano(mob/living/carbon/human/user, does_attack_animation = 0)
+	SendSignal(COMSIG_ATOM_HULK_ATTACK, user)
+	if(does_attack_animation)
+		user.changeNext_move(CLICK_CD_MELEE)
+		add_logs(user, src, "punched", "nanosuit strength mode")
+		user.do_attack_animation(src, ATTACK_EFFECT_SMASH)
+
+/obj/item/attack_nano(mob/living/carbon/human/user)
+	return 0
```
Comment (GuyonBroadway) [inline hippiestation/code/modules/clothing/spacesuits/nanosuit.dm]: return FALSE

File: hippiestation/code/modules/clothing/spacesuits/nanosuit.dm [hunk 10]
```diff
@@ -0,0 +1,934 @@
...
+
+/obj/effect/attack_nano(mob/living/carbon/human/user, does_attack_animation = 0)
+	return 0
```
Comment (GuyonBroadway) [inline hippiestation/code/modules/clothing/spacesuits/nanosuit.dm]: return FALSE

File: hippiestation/code/modules/clothing/spacesuits/nanosuit.dm [hunk 11]
```diff
@@ -0,0 +1,934 @@
...
+	new /obj/item/clothing/suit/space/hardsuit/nano(src)
+
+/obj/item/implant/explosive/disintegrate
+	name = "disintegration implant"
+	desc = "Ashes to ashes."
+	icon_state = "explosive"
+
+/obj/item/implant/explosive/disintegrate/activate(cause)
+	if(!cause || !imp_in || active)
+		return 0
```
Comment (GuyonBroadway) [inline hippiestation/code/modules/clothing/spacesuits/nanosuit.dm]: return FALSE

File: hippiestation/code/modules/clothing/spacesuits/nanosuit.dm [hunk 12]
```diff
@@ -0,0 +1,934 @@
...
+	if(cause == "action_button" && !popup)
+		popup = TRUE
+		var/response = alert(imp_in, "Are you sure you want to activate your [name]? This will cause you to vapourize!", "[name] Confirmation", "Yes", "No")
+		popup = FALSE
+		if(response == "No")
+			return 0
```
Comment (GuyonBroadway) [inline hippiestation/code/modules/clothing/spacesuits/nanosuit.dm]: return FALSE

File: hippiestation/code/modules/clothing/spacesuits/nanosuit.dm [hunk 13]
```diff
@@ -0,0 +1,934 @@
...
+
+	return 1
+
+/datum/martial_art/nano/harm_act(var/mob/living/carbon/human/A, var/mob/living/carbon/D)
+	add_logs(A, D, "punched")
+	var/picked_hit_type = pick("punches", "kicks")
+	var/bonus_damage = 10
+
+	if(D.IsKnockdown() || D.resting || D.lying)//we can hit ourselves
+		bonus_damage += 5
```
Comment (GuyonBroadway) [inline hippiestation/code/modules/clothing/spacesuits/nanosuit.dm:705]: Correct me if I am wrong here but won't this make the bonus damage grow by 5 every time? like so the bonus would go 5, 10, 15 etc?

**Review by GuyonBroadway — CHANGES_REQUESTED** [review]

File: hippiestation/code/modules/clothing/spacesuits/nanosuit.dm (hunk 13 above)
Comment (yoyobatty) [reply hippiestation/code/modules/clothing/spacesuits/nanosuit.dm:705]: no, the var/bonus_damage is reset to 10 every time the harm_act is called as it called every time so long as the proc returns true

File: hippiestation/code/modules/clothing/spacesuits/nanosuit.dm (hunk 2 above)
Comment (yoyobatty) [reply hippiestation/code/modules/clothing/spacesuits/nanosuit.dm]: yes, I will change that

File: hippiestation/code/modules/clothing/spacesuits/nanosuit.dm (hunk 1 above)
Comment (yoyobatty) [reply hippiestation/code/modules/clothing/spacesuits/nanosuit.dm:135]: yeah, that is the intention of this, it's designed to intercept all types of signals

**Comment by JohnGinnane** [comment]
@yoyobatty I haven't looked through the code (yet!) but is there a way for people to acquire this? Like 20TC item or something?

**Comment by GuyonBroadway** [comment]
@JohnGinnane Yeah its 20tc. Even has an announcement when someone uses it but I do feel it might be a bit OTT.

**Comment by JohnGinnane** [comment]
cool cool cool

**Comment by yoyobatty** [comment]
Feel free to test it out if you feel it is unbalanced. I think it consumes and recharges power at fair rates so you won't spam abilities. Also, Ion rifles zap the cell charge. For what it is it forces you to wear an outfit that gives you protective gear that disallows combining with other traitor items such as thermals or gloves of North Star. You also start with a no dismemberment trait which ensures that does not happen. To support that, the idea is that everything that comes with it is built into suit just like the game of origin. It's a literal full body suit.

**Comment by JohnGinnane** [comment]
I'm happy to merge as is and if it turns out to be OP we can make changes accordingly. This way we get the item in asap

**Comment by yoyobatty** [comment]
I'm pushing a new commit soon considering the changes suggested. I will run a test to make sure all is well and then go.

**Comment by JohnGinnane** [comment]
Need @HippieStation/head input pls))

**Comment by Pyko1** [comment]
Definitely should not be denied, but it sounds a bit op.

But no harm testing, I guess.

**Comment by Pyko1** [comment]
so +1 to see how much it needs to be tuned

**Review by GuyonBroadway — APPROVED** [review]

**Comment by Carbonhell** [comment]
A cautious +1 from me, but i think we will have to lock this to nukies, or nerf it, it sounds too op even for 20tcs

**Comment by yoyobatty** [comment]
@Carbonhell lets give it a test first and after a day or 2 if it seems unfair then I'll nerf it. Just let me remind you that the ion rifle is the suits worst enemy.

**Comment by GuyonBroadway** [comment]
has +2 and the code looks ok... here goes.

**Pull request merged**